### ISSUE TITLE
Unable to mmap RDWR log file: cannot allocate memory

### ISSUE THREAD

**Issue by GwynethLlewelyn** [dgraph-io/badger#246]
While working on a small memory footprint solution (in my particular case: running Go in a limited, shared environment under FastCGI), and requiring to trade off memory consumption _vs_ more file I/O, the overall solution is to use `Opt.TableLoadingMode = options.FileIO`.

However, this does not affect the value log, which is _always_ mmap'ed. Even using `Opt.ValueLogFileSize = 1048576` (the lowest value allowed), this fails under commit 273f40c6c24e1b188b68aeba1577de15e53fc015 (currently being tested under a customised version of Debian Linux over which I have no direct control) with the error `Unable to mmap RDWR log file: cannot allocate memory`.

I wonder if there is a way to either change the operations with the value log to also use file I/O instead of mmap'ed files, and/or reduce the `ValueLogFileSize` to even lower values.

I haven't separately tested Go's mmap library to very low values; my suspicion is that my current configuration doesn't even allow mmap'ed files altogether, but I need to check first to make sure.

**Comment by deepakjois** [comment]
@GwynethLlewelyn Could you please help me by answering the following questions:

* What is the amount of memory on the system you are testing?
* Is it 32-bit or 64-bit?
* Do let me know the effect of using standard mmap on that machine. You could create a simple test program to do that. Vary the value of `size` below and let me know what you observe

```Go
package main

import "os"
import "github.com/dgraph-io/badger/y"

func TestMmap(t *testing.T) {
	t.Log("Trying mmap")
	flags := os.O_RDWR | os.O_CREATE | os.O_EXCL
	f, err := os.OpenFile("test.md", flags, 0666)
	defer f.Close()
	y.Check(err)
	size := int64(500 * 1024 * 1024)
	t.Logf("Size is : %v", size)
	_, err = y.Mmap(f,false,size)
	if err != nil {
		t.Errorf("mmap failed with error: %v", err)
	}
}
```

**Comment by deepakjois** [comment]
Its been a while since we got an update here. Please feel free to reopen if you have anything new to add.

**Comment by GwynethLlewelyn** [comment]
My apologies for the lack of reply. In the past weeks I've been doing other things, and as I returned to Badger I saw that the whole interface changed... several times. So I took some time to get things working again just as before...

To answer your questions: the server is currently running Ubuntu 14.04.5 LTS under an Intel 64bit architecture, and while it has 32 GB of RAM, it's also running a watchdog which will kill any 'rogue' application consuming more than ~280 MBytes of available resident memory, something which my application will easily consume after importing ~583000 keys (in batches of 1000 at a time). Under Mac OS X High Sierra the same application, when importing 9 million key/value pairs, will easily reach 1 GB of resident memory. The runtime garbage collector is forcefully called after each batch is finished, which makes memory consumption not grow exponentially (rather the contrary, it tends to hover around a fixed maximum, something in the order of 600-700 MBytes or so).

Nevertheless, the `mmap` tests work fine:

```
$ go test -v
=== RUN   TestMmap
--- PASS: TestMmap (0.00s)
	gosl_test.go:10: Trying mmap
	gosl_test.go:18: Size is : 134217728
	gosl_test.go:18: Size is : 268435456
	gosl_test.go:18: Size is : 402653184
	gosl_test.go:18: Size is : 536870912
	gosl_test.go:18: Size is : 671088640
	gosl_test.go:18: Size is : 805306368
	gosl_test.go:18: Size is : 939524096
PASS
ok  	gosl-basics	0.005s
```

So I tried to use `options.MemoryMap' again, but I got the same issue: when importing key/value pais, the little Go application will consume more and more resident RAM until it steps over the ~280 MB mark and gets killed by the watchdog.

Basically what I need to understand is what options to pass when having a tiny memory footprint to make sure that memory consumption never reaches a certain limit (especially during importing!).

**Comment by deepakjois** [comment]
@GwynethLlewelyn Since Badger 1.0 has been released, we don’t intend to make any backward incompatible changes. So things should be much more stable now.

Before we released Badger 1.0, we used to have an option that could control whether we read the value log via mmap, or via standard file i/o. We removed the option and just defaulted to mmap. However, it seems that people could use that option. It might help your situation as well.

We will be putting back that option soon. I will update thi issue when that happens.

**Comment by deepakjois** [comment]
@GwynethLlewelyn #341 is now merged into master. It will be in the next release of Badger.

 Please try setting the `ValueLogLoadingMode` option to `options.FileIO` and let us know how it goes with memory consumption on your host. 

We also have a section in our README regarding memory usage. Please give that a read as well: https://github.com/dgraph-io/badger#memory-usage

Closing this for now.
